base/server/healthcheck/pki/server/healthcheck/meta/connectivity.py

```diff
@@ -0,0 +1,157 @@
+import logging
+
+from pki.server.healthcheck.meta.plugin import CSPlugin, registry
+from ipahealthcheck.core.plugin import Result, duration
+from ipahealthcheck.core import constants
+from pki.client import PKIConnection
+from pki.cert import CertClient
+from pki.systemcert import SystemCertClient
+
+logger = logging.getLogger()
+
+
+@registry
+class DogtagCACertsConnectivityCheck(CSPlugin):
+    """
+    Test basic CA connectivity by using cert-find to fetch a cert
+    """
+
+    @duration
+    def check(self):
+        if not self.instance.is_valid():
+            logging.debug('Invalid instance: %s', self.instance.name)
+            yield Result(self, constants.CRITICAL,
+                         msg='Invalid PKI instance: %s' % self.instance.name)
+            return
+
+        self.instance.load()
+
+        ca = self.instance.get_subsystem('ca')
+
+        if not ca:
+            logger.debug("No CA configured, skipping dogtag CA connectivity check")
+            return
+
+        try:
+            # Make a plain HTTP GET request to /ca/admin/ca/getStatus REST api end point
+            # and check if the CA is ready
+            if ca.is_ready():
+                logger.debug("CA instance is running")
+
+                # Make a plain HTTP GET to "find" one certificate, to test that
+                # the server is up AND is able to respond back
+                connection = PKIConnection(protocol='https',
+                                           hostname='localhost',
+                                           port='8443')
+
+                cert_client = CertClient(connection)
+                cert = cert_client.list_certs(size=1)
+                cert_info = cert.cert_data_info_list[0]
+                if cert_info:
+                    # All we care is whether the serial_number is not NONE
+                    if cert_info.serial_number:
+                        logger.info("Serial number of retrieved cert: %s", cert_info.serial_number)
+                        yield Result(self, constants.SUCCESS,
+                                     serial_number=cert_info.serial_number,
+                                     subject_dn=cert_info.subject_dn,
+                                     type=cert_info.type,
+                                     status=cert_info.status)
```

Comment:
Is it necessary to return the certificate info? IIUC the healthcheck tool only cares if the server is down. If the server is up, the returned data will not be used.

Reply:
I was just trying to provide as much details as possible, especially when this tool gets integrated into Insights